crates/ra_lsp_server/src/cargo_check.rs

```diff
@@ -0,0 +1,1316 @@
+        self.suggested_fix_collection.clear();
+
+        for uri in cleared_files {
+            task_send.send(CheckTask::Update(uri.clone())).unwrap();
+        }
+    }
+
+    pub fn diagnostics_for(&self, uri: &Url) -> Option<&[Diagnostic]> {
+        self.diagnostic_collection.get(uri).map(|d| d.as_slice())
+    }
+
+    pub fn fixes_for(&self, uri: &Url) -> Option<&[SuggestedFix]> {
+        self.suggested_fix_collection.get(uri).map(|d| d.as_slice())
+    }
+
+    fn add_diagnostic(&mut self, file_uri: Url, diagnostic: Diagnostic) {
+        let diagnostics = self.diagnostic_collection.entry(file_uri).or_default();
+
+        // If we're building multiple targets it's possible we've already seen this diagnostic
+        let is_duplicate = diagnostics.iter().any(|d| are_diagnostics_equal(d, &diagnostic));
+        if is_duplicate {
+            return;
+        }
+
+        diagnostics.push(diagnostic);
+    }
+
+    fn add_suggested_fix_for_diagnostic(
+        &mut self,
+        mut suggested_fix: SuggestedFix,
+        diagnostic: &Diagnostic,
+    ) {
+        let file_uri = suggested_fix.location.uri.clone();
+        let file_suggestions = self.suggested_fix_collection.entry(file_uri).or_default();
+
+        let existing_suggestion: Option<&mut SuggestedFix> =
+            file_suggestions.iter_mut().find(|s| s == &&suggested_fix);
+        if let Some(existing_suggestion) = existing_suggestion {
+            // The existing suggestion also applies to this new diagnostic
+            existing_suggestion.diagnostics.push(diagnostic.clone());
+        } else {
+            // We haven't seen this suggestion before
+            suggested_fix.diagnostics.push(diagnostic.clone());
+            file_suggestions.push(suggested_fix);
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum CheckTask {
+    /// Request a update of the given files diagnostics
+    Update(Url),
+
+    /// Request check progress notification to client
+    Status(WorkDoneProgress),
+}
+
+pub enum CheckCommand {
+    /// Request re-start of check thread
+    Update,
+}
+
+impl CheckWatcherState {
+    pub fn new(
+        options: Options,
+        workspace_root: PathBuf,
+        shared: Arc<RwLock<CheckWatcherSharedState>>,
+    ) -> CheckWatcherState {
+        let watcher = WatchThread::new(&options, &workspace_root);
+        CheckWatcherState {
+            options,
+            workspace_root,
+            running: false,
+            watcher,
+            last_update_req: None,
+            shared,
+        }
+    }
+
+    pub fn run(&mut self, task_send: &Sender<CheckTask>, cmd_recv: &Receiver<CheckCommand>) {
+        self.running = true;
+        while self.running {
+            select! {
+                recv(&cmd_recv) -> cmd => match cmd {
+                    Ok(cmd) => self.handle_command(cmd),
+                    Err(RecvError) => {
+                        // Command channel has closed, so shut down
+                        self.running = false;
```

Comment:
```suggestion
                        break;
```

and I think we won't need `self.running` field then?

Reply:
Yeah, that seems reasonable. I had an explicit shutdown command at some point, which is why `running` is there, but it didn't turn out necessary.

Reply:
Yeah, explicit shutdown is rarely necessary: dropping the sender usually works better